cachi2/core/package_managers/rpm/main.py

```diff
@@ -0,0 +1,301 @@
+import asyncio
+import hashlib
+import logging
+import shlex
+from os import PathLike
+from pathlib import Path
+from typing import Any, Union
+from urllib.parse import quote
+
+import yaml
+from pydantic import ValidationError
+
+from cachi2.core.config import get_config
+from cachi2.core.errors import PackageManagerError, PackageRejected
+from cachi2.core.models.input import Request
+from cachi2.core.models.output import RequestOutput
+from cachi2.core.models.sbom import Component, Property
+from cachi2.core.package_managers.general import async_download_files
+from cachi2.core.package_managers.rpm.redhat import RedhatRpmsLock
+from cachi2.core.rooted_path import RootedPath
+from cachi2.core.utils import run_cmd
+
+log = logging.getLogger(__name__)
+
+
+DEFAULT_LOCKFILE_NAME = "rpms.lock.yaml"
+DEFAULT_PACKAGE_DIR = "deps/rpm"
+
+# during the computing of file checksum read chunk of size 1 MB
+READ_CHUNK = 1048576
+
+
+def fetch_rpm_source(request: Request) -> RequestOutput:
+    """Process all the rpm source directories in a request."""
+    components: list[Component] = []
+    for package in request.rpm_packages:
+        path = request.source_dir.join_within_root(package.path)
+        components.extend(_resolve_rpm_project(path, request.output_dir))
+
+    return RequestOutput.from_obj_list(
+        components=components,
+        environment_variables=[],
+        project_files=[],
+    )
+
+
+def _resolve_rpm_project(source_dir: RootedPath, output_dir: RootedPath) -> list[Component]:
+    """
+    Process a request for a single RPM source directory.
+
+    Process the input lockfile, fetch packages and generate SBOM.
+    """
+    # Check the availability of the input lockfile.
+    if not source_dir.join_within_root(DEFAULT_LOCKFILE_NAME).path.exists():
+        raise PackageRejected(
+            f"RPM lockfile '{DEFAULT_LOCKFILE_NAME}' missing, refusing to continue.",
+            solution=(
+                "Make sure your repository has RPM lockfile '{DEFAULT_LOCKFILE_NAME}' checked in "
+                "to the repository."
+            ),
+        )
+
+    lockfile_name = source_dir.join_within_root(DEFAULT_LOCKFILE_NAME)
+    log.info(f"Reading RPM lockfile: {lockfile_name}")
+    with open(lockfile_name) as f:
+        try:
+            yaml_content = yaml.safe_load(f)
+        except Exception as e:
+            # log detail of the validation error
+            log.error(str(e))
```

Comment:
'msg' is not enough sometimes. Otherwise, we could end up with the message "Field required" without any context.
At least a 'loc' key needs to be considered.

Reply:
> msg' is not enough sometimes. Otherwise, we could end up with the message "Field required" without any context.
> At least a 'loc' key needs to be considered.

True. Makes sense, I was about to suggest something more readable, but that would end up with too much work, so 'loc' sounds like a good tradeoff.